### PR TITLE
SimulatedTest for HaltPlay

### LINKED ISSUE #1834: Implement SimulatedTest for HaltPlay
### Description of the task
Write a SimulatedTest for the HaltPlay

### Acceptance criteria

- [ ] a SimulatedTest for the HaltPlay

### PR DESCRIPTION
**Pull request by alinas2000**
### Description

Implemented a simulated test for Halt Play

### Testing Done

Passes bazel test //software/...

### Resolved Issues

Resolves #1834

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/test_util/test_util.cpp [hunk 1]
```diff
@@ -110,4 +110,15 @@ namespace TestUtil
...
+    void handleTestRobotLog(TbotsProto_RobotLog robot_log)
+    {
+        LOG(INFO) << "[TEST ROBOT " << robot_log.robot_id << "]["
+                  << TbotsProto::LogLevel_Name(
+                         static_cast<TbotsProto::LogLevel>(robot_log.log_level))
+                  << "]"
+                  << "[" << robot_log.file_name << ":" << robot_log.line_number
+                  << "]: " << robot_log.log_msg;
+    }
+
```
Comment (jonathanlew) [inline src/software/test_util/test_util.cpp]: I don't think this change is necessary? Is this from a bad merge?

File: src/software/simulated_tests/terminating_validation_functions/robot_halt_validation.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+#include "software/simulated_tests/validation/validation_function.h"
+#include "software/world/world.h"
+
+/**
+ * Checks if robots have stopped moving
+ * @param world_ptr the world pointer given by the simulator. Gets updated every tick
+ * @param yield yields control to the next routine (coroutines) with error message
+ */
+void robot_halt(std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield);
```
Comment (jonathanlew) [inline src/software/simulated_tests/terminating_validation_functions/robot_halt_validation.h]: camel case please

File: src/software/ai/hl/stp/play/halt_play_test.cpp [hunk 3]
```diff
@@ -33,10 +34,7 @@ TEST_F(HaltPlayTest, test_halt_play)
         // TODO: Implement proper validation
         // https://github.com/UBC-Thunderbots/Software/issues/1971
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/halt_play_test.cpp]: please remove this todo and initialize the robots with some velocities

**Review by jonathanlew — CHANGES_REQUESTED** [review]

**Review by jonathanlew — COMMENTED** [review]
just got a few comments

**Review by amrm3lm — COMMENTED** [review]
lgtm pending comments

File: src/software/simulated_tests/terminating_validation_functions/robot_halt_validation.h (hunk 2 above)
Comment (alinas2000) [reply src/software/simulated_tests/terminating_validation_functions/robot_halt_validation.h]: Fixed

File: src/software/ai/hl/stp/play/halt_play_test.cpp (hunk 3 above)
Comment (alinas2000) [reply src/software/ai/hl/stp/play/halt_play_test.cpp]: Created a new function to initialize velocities and used that in the test

File: src/software/test_util/test_util.cpp (hunk 1 above)
Comment (alinas2000) [reply src/software/test_util/test_util.cpp]: I believe so

Comment (jonathanlew) [reply src/software/test_util/test_util.cpp]: can you remove `handleTestRobotLog` then please?

Comment (alinas2000) [reply src/software/test_util/test_util.cpp]: done

File: src/software/test_util/test_util.h [hunk 4]
```diff
@@ -129,4 +129,15 @@ namespace TestUtil
      */
     std::vector<RobotStateWithId> createStationaryRobotStatesWithId(
         const std::vector<Point> &positions);
+
+    /**
+     * Creates a list of RobotStateWithId at given positions with input velocity, 0
+     * angular velocity and 0 orientation The id is set as the index in the list of
+     * positions
```
Comment (akhilveeraghanta) [inline src/software/test_util/test_util.h]: uber mega nit: missing a period after orientation

**Review by akhilveeraghanta — DISMISSED** [review]
nice work!

File: src/software/ai/hl/stp/play/halt_play_test.cpp [hunk 5]
```diff
@@ -30,13 +33,13 @@ TEST_F(HaltPlayTest, test_halt_play)
...
         // time to settle into position and be observed with the Visualizer
-        // TODO: Implement proper validation
-        // https://github.com/UBC-Thunderbots/Software/issues/1971
+
         [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
             while (world_ptr->getMostRecentTimestamp() < Timestamp::fromSeconds(9.5))
             {
                 yield("Timestamp not at 9.5s");
             }
+            robotHalt(world_ptr, yield);
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/halt_play_test.cpp]: waiting 9.5 seconds isn't necessary anymore and also it'd be good to check robotHalt over multiple ticks, i.e.
```suggestion
            for (unsigned i = 0; i < 1000; i++)
            {
                robotHalt(world_ptr, yield);
            }
```

**Review by jonathanlew — COMMENTED** [review]
just one more thing

File: src/software/test_util/test_util.h (hunk 4 above)
Comment (alinas2000) [reply src/software/test_util/test_util.h]: Fixed

File: src/software/ai/hl/stp/play/halt_play_test.cpp (hunk 5 above)
Comment (alinas2000) [reply src/software/ai/hl/stp/play/halt_play_test.cpp]: Fixed

**Review by jonathanlew — DISMISSED** [review]
LGTM!

**Review by akhilveeraghanta — COMMENTED** [review]
lgtm pending merge conflict!

**Review by jonathanlew — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**